third_party/move/move-compiler-v2/src/seqs_in_binop_checker.rs

```diff
@@ -0,0 +1,222 @@
+                            let rhs = &exps[1];
+                            let lhs_vars = lhs.free_vars_and_used_params(&param_symbols);
+                            let rhs_vars = rhs.free_vars_and_used_params(&param_symbols);
+                            let overlap = lhs_vars.intersection(&rhs_vars).next().is_some();
+                            if overlap
+                                || contains_control_flow_redirections(lhs)
+                                || contains_control_flow_redirections(rhs)
+                            {
+                                // Note: if needed, we can make this check even more precise by tracking
+                                // read and written variables, and checking only for read-write and
+                                // write-write conflicts.
+                                errors.push((binop_id, binop.clone(), seq_id));
+                            }
+                        }
+                    }
+                },
+                Sequence(id, seq)
+                    if seq.len() > 1 && !seq.iter().all(|exp| exp.is_ok_to_remove_from_code()) =>
+                {
+                    if let Some((binop_id, _)) = binop_stack.last() {
+                        // There is a non-trivial sequence within the context of a binary operation.
+                        // Non-trivial currently means that the sequence has more than one expression,
+                        // and not all those expressions are potentially side-effect free.
+                        // Note: if needed, we can implement a more precise check instead of reusing
+                        // `is_ok_to_remove_from_code` to track side-effect-free expressions.
+                        sequences.entry(*binop_id).or_insert(*id);
+                    }
+                },
+                _ => {},
+            }
+            true // continue traversal
+        };
+        def.visit_pre_post(&mut visitor);
+        let env = function.module_env.env;
+        for (binop_id, binop, seq_id) in errors {
+            let binop_loc = env.get_node_loc(binop_id);
+            let seq_loc = env.get_node_loc(seq_id);
+            let labels = vec![(seq_loc, "non-empty sequence".to_owned())];
+            let binop_as_str = binop.to_string_if_binop().expect("binop");
+            let notes = vec![
+                "To compile this code, either:".to_owned(),
+                "1. upgrade to language version 2.0 or above,".to_owned(),
+                "2. rewrite the code to remove sequences from directly within binary operations,"
+                    .to_owned(),
+                "   e.g., save intermediate results providing explicit order.".to_owned(),
+            ];
+            env.diag_with_primary_notes_and_labels(
+                Severity::Error,
+                &binop_loc,
+                &format!(
+                    "Non-empty sequence within the context of the binary operation `{}`",
```

Comment:
This error needs more work.  It would help to add something like

... is now disallowed due to unpredictable behavior in original Move compiler and language.

but this deserves careful attention.

Reply:
Sure. I'll incorporate your language suggestion here.

Reply:
Lets not use negative language about the old compiler in error messages. Better lets not talk at all about "original compiler" in error messages, in 1 year from now, no one knows what this is. 

Probably I would say something like 'Non-empty sequence as parameter of binary operator obscures program logic and is not supported by this compiler'.

We also should make clear that for solution (2), the semantics will be different, and the user should be sure to check that the program still does what is expected. "Upgrade to language version 2.0 or later which uses strict left-to-right evaluation order (be sure your program still does what you expect)"